File: algorithms/M/Maximum XOR With an Element From Array/Maximum XOR With an Element From Array.cpp
struct Trie {
    Trie *z;
    Trie *o;
    Trie() {
        z = NULL;
        o = NULL;
    }
};

class Solution {
public:
    
    Trie* insert(Trie* node, int k, int num) {
        if(node == NULL) {
            node = new Trie();
        }
        if(k < 0) return node;
        if(((num >> k) & 1) == 1) {
            node -> o = insert(node->o, k - 1, num);
        }
        else {
            node -> z = insert(node->z, k - 1, num);
        }
        return node;
    }
    
    int getmax(Trie* node, int k, int num) {
        if(k < 0) return 0;
        int res = 0;
        bool bit = ((num >> k) & 1);
        if(bit == 1) {
            if(node->z) {
                res = pow(2, k) + getmax(node->z, k - 1, num);
            }
            else {
                res = getmax(node->o, k - 1, num);
            }
        }
        else {
            if(node->o) {
                res = pow(2, k) + getmax(node->o, k - 1, num);
            }
            else {
                res = getmax(node->z, k - 1, num);
            }
        }
        return res;
    }
    
    vector<int> maximizeXor(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        int m = queries.size();
        vector<int> ans(m, -1);
        vector<pair<pair<int, int>, int> > q;
        for(int i = 0; i < m; i++) {
            q.push_back({{queries[i][0], queries[i][1]}, i});
        }
        sort(q.begin(), q.end(), [](pair<pair<int, int>, int> a, pair<pair<int, int>, int> b) {
            return a.first.second < b.first.second;
        });
        
        sort(nums.begin(), nums.end());
        
        int j = 0;
        
        Trie *root = new Trie();
        
        for(int i = 0; i < m; i++) {
            pair<pair<int, int>, int> p = q[i];
            int num = p.first.first;
            int lim = p.first.second;
            int idx = p.second;
            
            while(j < n && nums[j] <= lim) {
                root = insert(root, 31, nums[j]);
                j++;
            }
            
            if(j > 0) {
                ans[idx] = getmax(root, 31, num);
            }
        }
        
        return ans;
    }
};

File: scripts/algorithms/M/Maximum XOR With an Element From Array/Maximum XOR With an Element From Array.java
class Solution {
    class QueryComparator implements Comparator<int[]> {
        public int compare(int[] a, int[] b) {
            return Integer.compare(a[1], b[1]);
        }
    }
    
    class Node {
        Node zero;
        Node one;
        
        public Node() {
            this.zero = null;
            this.one = null;
        }
    }
    
    public int[] maximizeXor(int[] nums, int[][] queries) {
        Arrays.sort(nums);
        
        int len = queries.length;
        int[][] queryWithIndex = new int[len][3];
        for(int i = 0; i < len; i++) {
            queryWithIndex[i][0] = queries[i][0];
            queryWithIndex[i][1] = queries[i][1];
            queryWithIndex[i][2] = i;
        }
        Arrays.sort(queryWithIndex, new QueryComparator());
        
        int numId = 0;
        int[] ans = new int[len];
        
        Node root = new Node();
        for(int i = 0; i < len; i++) {
            while(numId < nums.length && nums[numId] <= queryWithIndex[i][1]) {
                addNumToTree(nums[numId], root);
                numId++;
            }
            
            ans[queryWithIndex[i][2]] = maxXOR(queryWithIndex[i][0], root);
        }
        
        return ans;
    }
    
    private void addNumToTree(int num, Node node) {
        for(int i = 31; i >= 0; i--) {
            int digit = (num >> i) & 1;
            if (digit == 1) {
                if (node.one == null) {
                    node.one = new Node();
                }
                node = node.one;
            } else {
                if (node.zero == null) {
                    node.zero = new Node();
                }
                node = node.zero;
            }
        }
    }
    
    private int maxXOR(int num, Node node) {
        if (node.one == null && node.zero == null) {
            return -1;
        }

        int ans = 0;
        for(int i = 31; i >= 0 && node != null; i--) {
            int digit = (num >> i) & 1;
            if (digit == 1) {
                if (node.zero != null) {
                    ans += (1 << i);
                    node = node.zero;
                } else {
                    node = node.one;
                }
            } else {
                if (node.one != null) {
                    ans += (1 << i);
                    node = node.one;
                } else {
                    node = node.zero;
                }
            }
        }
        
        return ans;
    }
}
